File: pc/plugins/remainder.py
import re
from core.plugin import Plugin


class RemainderPlugin(Plugin):
    def translate(self, line):
        match = re.fullmatch(r'余り\s+(.+?)\s+(.+)', line)

        if not match:
            return None

        name = match.group(1)
        value = match.group(2)

        return f"{name} %= {value}"

File: web/plugins/remainder.js
import { WebPlugin } from "../engine/plugin.js";

export class RemainderPlugin extends WebPlugin {
    constructor() {
        super("余り");
    }

    async execute(line, context) {
        const match = line.match(/^余り\s+(.+?)\s+(.+)$/);

        if (!match) {
            throw new Error(`余りの書式が正しくありません: ${line}`);
        }

        const [, target, expression] = match;

        const value = context.parser.parseNumber(
            expression,
            context.runtime
        );

        if (value === 0) {
            throw new Error("0で割ることはできません。");
        }

        if (context.runtime.exists(target)) {
            const current = context.parser.parseNumber(
                target,
                context.runtime
            );

            context.runtime.set(target, current % value);
            return;
        }

        const current = context.parser.parseNumber(
            target,
            context.runtime
        );

        context.runtime.write(current % value);
    }
}
